File: tobiasHe/assignments/slidingwindow/lc53/lc53.js
function max_subarray_size_k(k, arr) {
	// TODO: Write your code here

	//3, [ 2, 1, 5, 1, 3, 2 ])

	let maxSum = 0;
	let currentSum = 0;
	let windowStart = 0;
	let windowEnd = 0;

	for (let i = 0; i < arr.length; i++) {
		currentSum += arr[i];
		windowEnd++;
		if (windowEnd > k - 1) {
			maxSum = Math.max(maxSum, currentSum);
			//subtract element
			currentSum = currentSum - arr[windowStart];
			windowStart++;
		}
	}
	return maxSum;

	// let maxSum = 0,
	// 	windowSum = 0,
	// 	windowStart = 0;

	// for (window_end = 0; window_end < arr.length; window_end++) {
	// 	// add the next element
	// 	windowSum += arr[window_end];
	// 	// slide the window, we don't need to slide if we've not hit the required window size of 'k'
	// 	if (window_end >= k - 1) {
	// 		maxSum = Math.max(maxSum, windowSum);
	// 		// subtract the element going out
	// 		windowSum -= arr[windowStart];
	// 		// slide the window ahead
	// 		windowStart += 1;
	// 	}
	// }
	// return maxSum;
}

console.log(`Maximum sum of a subarray of size K: ${max_subarray_size_k(3, [ 2, 1, 5, 1, 3, 2 ])}`); //9
console.log(`Maximum sum of a subarray of size K: ${max_subarray_size_k(2, [ 2, 3, 4, 1, 5 ])}`); //7

// bruteforce/naive approach
// -----
// let maxSum = 0,
// windowSum = 0;

// for (i = 0; i < arr.length - k + 1; i++) {
//   windowSum = 0;
//   for (j = i; j < i + k; j++) {
//     windowSum += arr[j];
//   }
//   maxSum = Math.max(maxSum, windowSum);
// }
// return maxSum;

// better approach
// -----
// let maxSum = 0,
// windowSum = 0,
// windowStart = 0;

// for (window_end = 0; window_end < arr.length; window_end++) {
//   // add the next element
//   windowSum += arr[window_end];
//   // slide the window, we don't need to slide if we've not hit the required window size of 'k'
//   if (window_end >= k - 1) {
//     maxSum = Math.max(maxSum, windowSum);
//     // subtract the element going out
//     windowSum -= arr[windowStart];
//     // slide the window ahead
//     windowStart += 1;
//   }
// }
// return maxSum;

File: mariannaFervenza/assignments/slidingwindow/lc53/lc53.cpp
using namespace std;

#include <iostream>
#include <vector>

class MaxSumSubArrayOfSizeK {
 public:
  static int MaxSumSubArray(int k, const vector<int>& arr) {
    int maxSum = 0
    // Write your code here
    return maxSum;
  }
};

int main(int argc, char* argv[]) {
  cout << "Maximum sum of a subarray of size K: "
       << MaxSumSubArrayOfSizeK::MaxSumSubArray(3, vector<int>{2, 1, 5, 1, 3, 2}) << endl;
  cout << "Maximum sum of a subarray of size K: "
       << MaxSumSubArrayOfSizeK::MaxSumSubArray(2, vector<int>{2, 3, 4, 1, 5}) << endl;
}








// bruteforce/naive approach
// -----
  //   int maxSum = 0, windowSum;
  //   for (int i = 0; i <= arr.size() - k; i++) {
  //     windowSum = 0;
  //     for (int j = i; j < i + k; j++) {
  //       windowSum += arr[j];
  //     }
  //     maxSum = max(maxSum, windowSum);
  //   }

  //   return maxSum;
  // }


// a better approach
// -----
  //   int windowSum = 0, maxSum = 0;
  //   int windowStart = 0;
  //   for (int windowEnd = 0; windowEnd < arr.size(); windowEnd++) {
  //     windowSum += arr[windowEnd];  // add the next element
  //     // slide the window, we don't need to slide if we've not hit the required window size of 'k'
  //     if (windowEnd >= k - 1) {
  //       maxSum = max(maxSum, windowSum);
  //       windowSum -= arr[windowStart];  // subtract the element going out
  //       windowStart++;                  // slide the window ahead
  //     }
  //   }

  //   return maxSum;
  // }
